pkg/network/ebpf/c/http.h

```diff
@@ -167,51 +142,98 @@ static __always_inline void http_parse_data(char *p, http_packet_t *packet_type,
+
+    if (!must_flush) {
+        return NULL;
+    }
+
+    u32 cpu = bpf_get_smp_processor_id();
+    http_batch_state_t *batch_state = bpf_map_lookup_elem(&http_batch_state, &cpu);
+    if (batch_state == NULL) {
+        return NULL;
+    }
+
+    __builtin_memcpy(&batch_state->scratch_tx, http, sizeof(http_transaction_t));
+    return &batch_state->scratch_tx;
+}
+
+static __always_inline bool http_closed(http_transaction_t *http, skb_info_t *skb_info, u16 pre_norm_src_port) {
+    return (skb_info && skb_info->tcp_flags&TCPHDR_FIN &&
+            http->owned_by_src_port == pre_norm_src_port);
+}
+
+static __always_inline int http_process(http_transaction_t *http_stack, skb_info_t *skb_info) {
+    char *buffer = (char *)http_stack->request_fragment;
     http_packet_t packet_type = HTTP_PACKET_UNKNOWN;
     http_method_t method = HTTP_METHOD_UNKNOWN;
     http_parse_data(buffer, &packet_type, &method);
-    http_transaction_t *http = NULL;
 
-    http_transaction_t new_entry = { 0 };
-    new_entry.owned_by_src_port = src_port;
+    http_transaction_t *http = http_fetch_state(http_stack, skb_info, packet_type);
+    if (http == NULL) {
+        return 0;
+    }
 
-    switch(packet_type) {
-    case HTTP_REQUEST:
-        bpf_map_update_elem(&http_in_flight, &skb_info->tup, &new_entry, BPF_NOEXIST);
-        http = bpf_map_lookup_elem(&http_in_flight, &skb_info->tup);
-        if (http == NULL || http->owned_by_src_port != src_port) {
-            return 0;
-        }
-        http_begin_request(http, method, buffer, &skb_info->tup);
-        break;
-    case HTTP_RESPONSE:
-        bpf_map_update_elem(&http_in_flight, &skb_info->tup, &new_entry, BPF_NOEXIST);
-        http = bpf_map_lookup_elem(&http_in_flight, &skb_info->tup);
-        if (http == NULL) {
-            return 0;
-        }
+    http_transaction_t *to_flush = http_should_flush_previous_state(http, packet_type);
+    if (packet_type == HTTP_REQUEST) {
+        http_begin_request(http, method, buffer);
+    } else if (packet_type == HTTP_RESPONSE) {
         http_begin_response(http, buffer);
-        break;
-    default:
-        // We're either in the middle of either a request or response
-        http = bpf_map_lookup_elem(&http_in_flight, &skb_info->tup);
-        if (http == NULL) {
-            return 0;
-        }
     }
 
-    http->tags |= tags;
-
     // If we have a (L7/application-layer) payload we want to update the response_last_seen
     // This is to prevent things such as a keep-alive adding up to the transaction latency
     if (buffer[0] != 0) {
         http->response_last_seen = bpf_ktime_get_ns();
     }
 
-    if (skb_info->tcp_flags & TCPHDR_FIN && http->owned_by_src_port == src_port) {
-        http_enqueue(http, &skb_info->tup);
-        bpf_map_delete_elem(&http_in_flight, &skb_info->tup);
+    bool conn_closed = http_closed(http, skb_info, http_stack->owned_by_src_port);
+    if (conn_closed) {
+        to_flush = http;
```

Comment:
Can you explain the relationship of `to_flush` and `http` here? I'm not following why you would overwrite the value here. It seems `to_flush` is meant to represent previous state that needs to be flushed, and thus would be lost when overwritten.

Reply:
There is no real overriding there in the sense that when `conn_closed` is `true`, `to_flush` is `NULL`. I don't like how this code reads TBH. This is how I wish I could have written this code:

```c
if (conn_closed) {
  http_enqueue(http);
}
```
But this would mean that we would inline `http_enqueue` twice (which is super a beefy function) and there will be less room (in terms of instructions) for the HTTP path increase 😞